File: boj/sohan/1991.py
N = int(input())
trees = [input().split(' ') for _ in range(N)]

def findidx(tree, c):
    for i in range(len(tree)):
        if tree[i][0] == c:
            return i
    return -1

def preorder(node):
    if not node:
        return ;
    print(node[0], end='')
    left = findidx(trees, node[1])
    if left  == -1:
        leftnode = 0
    else:
        leftnode = trees[left]
    right = findidx(trees, node[2])
    if right == -1:
        rightnode = 0
    else:
        rightnode = trees[right]

    preorder(leftnode)
    preorder(rightnode)

def inorder(node):
    if not node:
        return ;
    left = findidx(trees, node[1])
    if left  == -1:
        leftnode = 0
    else:
        leftnode = trees[left]
    right = findidx(trees, node[2])
    if right == -1:
        rightnode = 0
    else:
        rightnode = trees[right]

    inorder(leftnode)
    print(node[0], end='')
    inorder(rightnode)

def postorder(node):
    if not node:
        return ;
    left = findidx(trees, node[1])
    if left  == -1:
        leftnode = 0
    else:
        leftnode = trees[left]
    right = findidx(trees, node[2])
    if right == -1:
        rightnode = 0
    else:
        rightnode = trees[right]

    postorder(leftnode)
    postorder(rightnode)
    print(node[0], end='')

preorder(trees[0])
print("")
inorder(trees[0])
print("")
postorder(trees[0])

File: boj/wmo/1991.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct s_tree{
	char data;
	struct s_tree *left;
	struct s_tree *right;
}t_tree;

int n;

void insert_node(t_tree *tree, char p, char l, char r) {
	if (l != '.')
		tree[p].left = &tree[l];
	if (r != '.')
		tree[p].right = &tree[r];
}

void preorder(t_tree *t) {
	if (t == 0)
		return ;
	printf("%c", t->data);
	preorder(t->left);
	preorder(t->right);
}

void inorder(t_tree *t) {
	if (t == 0)
		return ;
	inorder(t->left);
	printf("%c", t->data);
	inorder(t->right);
}

void postorder(t_tree *t) {
	if (t == 0)
		return ;
	postorder(t->left);
	postorder(t->right);
	printf("%c", t->data);
}

void free_data(t_tree *t) {
	if (t == 0)
		return ;
	free_data(t->left);
	free_data(t->right);
	free(t);
}

int main() {

	scanf("%d", &n);
	while (getchar() != '\n');

	t_tree *tree;
	char a, b, c;
	a = b = c = 65;

	tree = malloc(sizeof(t_tree) * 100);
	memset(tree, 0, sizeof(t_tree) * 100);
	for (int i = 65; i < 100; i++) 
		tree[i].data = i;
	while (n--) {
		scanf("\n%c %c %c", &a, &b, &c);
		insert_node(tree, a, b, c);
	}
	preorder(&tree[65]);
	printf("\n");
	inorder(&tree[65]);
	printf("\n");
	postorder(&tree[65]);
	free(tree);
	return 0;
}
